diff --git a/src/alloc.c b/src/alloc.c
index 10e85d6..b9f2121 100644
--- a/src/alloc.c
+++ b/src/alloc.c
@@ -804,8 +804,9 @@ typedef void (*std_new_handler_t)(void);
 #if (defined(__GNUC__) || (defined(__clang__) && !defined(_MSC_VER)))  // exclude clang-cl, see issue #631
-std_new_handler_t __attribute__((weak)) _ZSt15get_new_handlerv(void) {
-  return NULL;
-}
+extern std_new_handler_t _ZSt15get_new_handlerv(void) __attribute__((weak));
 static std_new_handler_t mi_get_new_handler(void) {
-  return _ZSt15get_new_handlerv();
+  if (_ZSt15get_new_handlerv == NULL) {
+    return NULL;
+  }
+  return _ZSt15get_new_handlerv();
 }
 #else
 // note: on windows we could dynamically link to `?get_new_handler@std@@YAP6AXXZXZ`.
